python/ray/train/utils.py

```diff
@@ -31,19 +33,46 @@ def check_for_failure(remote_values: List[ObjectRef]) -> Tuple[bool, List[int]]:
     """
     unfinished = remote_values.copy()
     dead_worker_indexes = []  # Store the indexes of the failed workers.
+
+    at_least_one_failed_worker = False
     while len(unfinished) > 0:
-        finished, unfinished = ray.wait(unfinished)
+        # Once a worker has failed, we add a timeout for getting the
+        # results for the remaining workers.
+        # This is to avoid situations where the remaining workers
+        # are alive, but hanging on collective calls because other
+        # workers have failed.
```

Comment:
Can this be generalized? Is this exclusive to collective calls or are there other situations this would help as well?

Reply:
Any situation where some workers dying cause other workers to hang.

Reply:
Updated the comment to reflect this